accepted/00000-maintenance-windows.md

```diff
@@ -0,0 +1,390 @@
+  44 | kickstart_host.schedule_virt_host_pkg_install   | Schedule a package install of host specific functionality.                                         | N
+  45 | kickstart_guest.schedule_virt_guest_pkg_install | Schedule a package install of guest specific functionality.                                        | N
+  46 | kickstart_host.add_tools_channel                | Subscribes a server to the Spacewalk Tools channel associated with its base channel.               | N
+  47 | kickstart_guest.add_tools_channel               | Subscribes a virtualization guest to the Spacewalk Tools channel associated with its base channel. | N
+  48 | virt.setVCPUs                                   | Sets the Vcpu usage for a virtual domain.                                                          | Y
+  49 | proxy.deactivate                                | Deactivate Proxy                                                                                   | N
+  50 | scap.xccdf_eval                                 | OpenSCAP xccdf scanning                                                                            | Y ?
+  51 | clientcert.update_client_cert                   | Update Client Certificate                                                                          | Y
+ 500 | image.deploy                                    | Deploy an image to a virtual host.                                                                 | N
+ 501 | distupgrade.upgrade                             | Service Pack Migration                                                                             | Y
+ 502 | packages.setLocks                               | Lock packages                                                                                      | N
+ 503 | states.apply                                    | Apply states                                                                                       | Y
+ 504 | image.build                                     | Build an Image Profile                                                                             | N
+ 505 | image.inspect                                   | Inspect an Image                                                                                   | N
+ 506 | channels.subscribe                              | Subscribe to channels                                                                              | Y ?
+
+(That list is currently the full action list. Not all actions are still in use)
+
+
+### Schedule an Action
+
+When a system is assigned to a maintenance schedule, actions which are flagged as `maintenance_mode_only` should be executed only within a maintenance window.
+In the User Interface we should show in these cases not the datepicker, but a simple Combo box with the Date and Time of the next 20 Maintenance Windows.
+All actions will be scheduled at the begin of the Maintenance Window.
```

Comment:
If we constrain all `maintainance_mode_only` Actions to be scheduled at the beginning of a Maintenance Window, does it make sense to talk about Maintenance Windows at all? Is this constraint expected to be permanent, or to be lifted in a successive iteration?

If it is expected to be permanent, wouldn't it be easier to "rename" this whole concept to "Maintenance Start Times" and ignore lengths altogether?

(the only difference would be to prevent out-of-window scheduling via XMLRPC - arguably though if a user already uses XMLRPC he might have a scheduling already)

Reply:
> If we constrain all `maintainance_mode_only` Actions to be scheduled at the beginning of a Maintenance Window, does it make sense to talk about Maintenance Windows at all? Is this constraint expected to be permanent, or to be lifted in a successive iteration?

I don' think we should schedule everything at the beginning of the maintenance window, at least if there are multiple actions to be scheduled, i. e. do not overlap actions.

> If it is expected to be permanent, wouldn't it be easier to "rename" this whole concept to "Maintenance Start Times" and ignore lengths altogether?
> 
> (the only difference would be to prevent out-of-window scheduling via XMLRPC - arguably though if a user already uses XMLRPC he might have a scheduling already)

Maintenance windows are a well-established concept in the ITSM world, I don't think there's a need to redefine concepts or try to outsmart sysadmins.

In the real world, it works like this:

* "as my maintenance window is from 2 AM to 6 AM on Saturday and Sunday night, I will plan to start this maintenance procedure at 3 AM on Saturday and I [human] have estimated this will be finished at 4 AM, because I have done my testing in TEST and PRE [which in theory is identical to PRO] environments"
* Why 3 AM instead of 2 AM, which is the time the maintenance window starts? Because it turns out this week there is some other procedure at 2 AM, at the beginning of the maintenance window. It does not even need to be for this server, but given that a human operator is always supervising or executing changes, a procedure on server A at 2 AM may prevent another procedure on server B at 2 AM due to lack of humans
* If for some reason (human or machine) the procedure does not finish on time, there are two possibilities:
  * Rollback is possible and/or desirable: a final "go/no-go" decision is made at e. g. 5 AM (whatever is enough to cancel the procedure and perform the rollback)
  * Rollback is not possible and/or undesirable: let the procedure finish at e. g. 7 AM, and register an out-of-window event (incident) in the ITSM tool (e. g. ServiceNow or BMC), calculate SLIs/SLAs, etc

Reply:
> I don' think we should schedule everything at the beginning of the maintenance window, at least if there are multiple actions to be scheduled, i. e. do not overlap actions to be executed on the same client.

Note: both the traditional and the Salt stack queue execution of Actions if they were to be concurrently on the same client.

Reply:
Subscribe to (software) channels should always be possible.

Reply:
It seems there is a preference to make it at least possible to schedule an action later in the maintenance window. I think I have a good idea how to make it easy possible.

We would implement a combobox with the next 20 maintenance windows and additionally an entry field with "Offset in Minutes" which has the default `0`. Using the offset, you can schedule an action 60 Minutes after the maintenance window starts. I also think remember an offset is easier than an exact time.

I will add this to the RFC

Reply:
> Subscribe to (software) channels should always be possible.

Will add this to the unresolved questions and we decide in the afternoon.

Reply:
Just a reminder: changing software channel subscriptions for minions might imply product package installation!

Reply:
But only when a highstate is called. (Traditional clients are different, but they are somehow excluded when thinking about this feature)

Reply:
Right, I forgot about that. Should not be a worry then.